river/Cursor.zig

```diff
@@ -802,26 +935,92 @@ fn processMotion(self: *Self, device: *wlr.InputDevice, time: u32, delta_x: f64,
             data.delta_y = dy - @trunc(dy);
 
             const border_width = if (data.view.draw_borders) server.config.border_width else 0;
-
-            // Set width/height of view, clamp to view size constraints and output dimensions
             const box = &data.view.pending.box;
-            box.width = @intCast(u32, math.max(0, @intCast(i32, box.width) + @floatToInt(i32, dx)));
-            box.height = @intCast(u32, math.max(0, @intCast(i32, box.height) + @floatToInt(i32, dy)));
+            const constraints = data.view.getConstraints();
+
+            // Avoid potential integer overflow. No reasonable client will have
+            // a minimum size even remotely as large, so this should not be
+            // interfering with well behaved toplevels, but it prevents bad view
+            // constraints from crashing us. A max size larger than this is
+            // pretty common however as most views do not have a max size, which
+            // means they have max_width = max_height = maxInt(u32).
+            const max = math.maxInt(i32) / 2;
+            const min_width = @intCast(i32, math.max(View.min_size, math.min(max, constraints.min_width)));
+            const min_height = @intCast(i32, math.max(View.min_size, math.min(max, constraints.min_height)));
+            const max_width = @intCast(i32, math.min(max, constraints.max_width));
+            const max_height = @intCast(i32, math.min(max, constraints.max_height));
+
+            // Using two points instead of a box makes this logic a bit more
+            // readable for humans. The compiler should optimize the conversion out.
```

Comment:
We really don't care if it does, this will be plenty fast either way.
```suggestion
            // Using two points instead of a box makes this logic a bit more
            // readable for humans.
```